File: output_repair/claude-opus/sample_1/repair_2/c/c/Stack/filter.c
#include "stdlib.h"

struct node
{
    struct node *next;
    int value;
};

struct stack
{
    struct node *head;
};

/*@
predicate nodes(struct node *n;) =
    n == 0 ? emp : n->next |-> ?next &*& n->value |-> _ &*& malloc_block_node(n) &*& nodes(next);

predicate stack(struct stack *s;) =
    s->head |-> ?head &*& malloc_block_stack(s) &*& nodes(head);
@*/

struct stack *create_stack()
//@ requires true;
//@ ensures stack(result);
{
    struct stack *stack = malloc(sizeof(struct stack));
    if (stack == 0)
    {
        abort();
    }
    stack->head = 0;
    //@ close nodes(0);
    //@ close stack(stack);
    return stack;
}

void stack_push(struct stack *stack, int value)
//@ requires stack(stack);
//@ ensures stack(stack);
{
    //@ open stack(stack);
    struct node *n = malloc(sizeof(struct node));
    if (n == 0)
    {
        abort();
    }
    n->next = stack->head;
    n->value = value;
    stack->head = n;
    //@ close nodes(n);
    //@ close stack(stack);
}

int stack_pop(struct stack *stack)
//@ requires stack(stack);
//@ ensures stack(stack);
{
    //@ open stack(stack);
    //@ open nodes(stack->head);
    struct node *head = stack->head;

    int result = head->value;
    stack->head = head->next;
    free(head);
    //@ close stack(stack);
    return result;
}

typedef bool int_predicate(int x);

struct node *nodes_filter(struct node *n, int_predicate *p)
//@ requires nodes(n);
//@ ensures nodes(result);
{
    //@ open nodes(n);
    if (n == 0)
    {
        //@ close nodes(0);
        return 0;
    }
    else
    {
        bool keep = p(n->value);
        if (keep)
        {
            struct node *next = nodes_filter(n->next, p);

            n->next = next;
            //@ close nodes(n);
            return n;
        }
        else
        {
            struct node *next = n->next;
            free(n);
            struct node *result = nodes_filter(next, p);
            return result;
        }
    }
}

void stack_filter(struct stack *stack, int_predicate *p)
//@ requires stack(stack);
//@ ensures stack(stack);
{
    //@ open stack(stack);
    struct node *head = nodes_filter(stack->head, p);

    stack->head = head;
    //@ close stack(stack);
}

void nodes_dispose(struct node *n)
//@ requires nodes(n);
//@ ensures true;
{
    //@ open nodes(n);
    if (n != 0)
    {
        nodes_dispose(n->next);
        free(n);
    }
}

void stack_dispose(struct stack *stack)
//@ requires stack(stack);
//@ ensures true;
{
    //@ open stack(stack);
    nodes_dispose(stack->head);
    free(stack);
}

bool neq_20(int x)
//@ requires true;
//@ ensures true;
{
    return x != 20;
}

int main()
//@ requires true;
//@ ensures true;
{
    struct stack *s = create_stack();
    stack_push(s, 10);
    stack_push(s, 20);
    stack_push(s, 30);
    stack_filter(s, neq_20);
    stack_dispose(s);
    return 0;
}

File: output_log/claude-opus/sample_5/rust/claude-opus/rust/purely_unsafe/solutions/filter.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

struct Node {
    next: *mut Node,
    value: i32,
}

struct Stack {
    head: *mut Node,
}

type I32Predicate = unsafe fn(i32) -> bool;

predicate nodes(n: *mut Node;) = 
    n != std::ptr::null_mut() ?
        (*n).next |-> ?next &*& (*n).value |-> ?v &*& malloc_block_Node(n) &*& nodes(next)
        : emp;

predicate stack(s: *mut Stack;) = 
    (*s).head |-> ?h &*& malloc_block_Stack(s) &*& nodes(h);

predicate_ctor predicate_req(p: I32Predicate)(v: int) =
    [_]is_I32Predicate(p) &*& // owns permission to call p
    true;

predicate_ctor predicate_ens(p: I32Predicate)(v: int; r: bool) =
    true;

#[requires(nodes(n) &*& [_]is_I32Predicate(p))]
#[ensures(nodes(result))]
unsafe fn filter_nodes(n: *mut Node, p: I32Predicate) -> *mut Node {
    if n.is_null() {
        std::ptr::null_mut()
    } else {
        let v = (*n).value;
        let next = (*n).next;
        predicate_req(p)(v);
        let keep = p(v);
        predicate_ens(p)(v, keep);
        if keep {
            let next_filtered = filter_nodes(next, p);
            (*n).next = next_filtered;
            n
        } else {
            dealloc(n as *mut u8, Layout::new::<Node>());
            filter_nodes(next, p)
        }
    }
}

#[requires(nodes(n))]
#[ensures(emp)]
unsafe fn dispose_nodes(n: *mut Node) {
    if !n.is_null() {
        let next = (*n).next;
        dispose_nodes(next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {
    #[ensures(stack(result))]
    unsafe fn create() -> *mut Stack {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();
        stack
    }

    #[requires(stack(stack))]
    #[ensures(stack(stack))]
    unsafe fn push(stack: *mut Stack, value: i32) {
        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }

        (*n).next = (*stack).head;
        (*n).value = value;
        (*stack).head = n;
    }

    #[requires(stack(stack))]
    #[ensures(stack(stack))]
    unsafe fn pop(stack: *mut Stack) -> i32 {
        let head = (*stack).head;
        // must own nodes(head)
        // head not null, otherwise popping empty stack not allowed
        open stack(stack);
        assume(head != std::ptr::null_mut());
        open nodes(head);

        let result = (*head).value;
        (*stack).head = (*head).next;
        close nodes((*head).next);
        dealloc(head as *mut u8, Layout::new::<Node>());
        close stack(stack);
        result
    }

    #[requires(stack(stack) &*& [_]is_I32Predicate(p))]
    #[ensures(stack(stack))]
    unsafe fn filter(stack: *mut Stack, p: I32Predicate) {
        let head = filter_nodes((*stack).head, p);
        (*stack).head = head;
    }

    #[requires(stack(stack))]
    #[ensures(emp)]
    unsafe fn dispose(stack: *mut Stack) {
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }
}

#[ensures(true)]
unsafe fn neq_20(x: i32) -> bool {
    x != 20
}

fn main() {
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);

        Stack::filter(s, neq_20);
        Stack::dispose(s);
    }
}